andela/ah-sealteam
# [Feature #161278918] Users should be able to follow each other

### What does this PR do?
- Create Friends app to handle following/unfollowing between users
### Description of Task to be completed?
- Add Friend model class
- Add the CustomSerializer to verify the usernames
- Add the view classes
- Add the routes
### Functionality
- A user can follow and be followed by other users
- A user can view users he/she is following.(Available for other users)
- A user can view users following him/her.(Available for other users)
### How should this be manually tested
- clone the Repo 'git clone https://github.com/andela/ah-sealteam.git'
- Install Virtualenv `pip install virtualenv`
- Create the virtual environment `virtualenv venv`
- Activate the virtualenv `source venv/bin/activate`
- navigate to project folder `cd ah-sealteam`
- Install the dependencies `pip install -r requirements.txt`
- Run the migrations `python manage.py makemigrations && python manage.py migrate`
<img width="1043" alt="screen shot 2018-11-07 at 19 36 24" src="https://user-images.githubusercontent.com/35170812/48145717-7f2e2c80-e2c4-11e8-9bf4-94cc9bc814d2.png">

- Register a user via the endpoint `http://127.0.0.1:8000/api/users`. This also logins the user returning a token. (Register multiple users).
<img width="885" alt="screen shot 2018-11-07 at 19 10 58" src="https://user-images.githubusercontent.com/35170812/48143930-ef3ab380-e2c0-11e8-82a7-7039f845c42c.png">`
- Follow one of the users using the endpoint POST `http://127.0.0.1:8000/api/profiles/USERNAME/follow`. The response returns the profile of the user followed
<img width="905" alt="screen shot 2018-11-07 at 19 13 29" src="https://user-images.githubusercontent.com/35170812/48144077-42146b00-e2c1-11e8-81a2-5948850fb10a.png">
- Check the users following that user GET `http://127.0.0.1:8000/api/profiles/USERNAME/followers`
<img width="885" alt="screen shot 2018-11-07 at 19 16 02" src="https://user-images.githubusercontent.com/35170812/48144275-a2a3a800-e2c1-11e8-9530-874407a79c05.png">
- Unfollow the user DELETE `http://127.0.0.1:8000/api/profiles/USERNAME/unfollow`
The response returns the profile of the unfollowed user
<img width="889" alt="screen shot 2018-11-07 at 19 18 03" src="https://user-images.githubusercontent.com/35170812/48144413-e1d1f900-e2c1-11e8-9251-49dee04e3b7b.png">
- Follow other users and check the users you're following GET `http://127.0.0.1:8000/api/profiles/USERNAME/following`
<img width="911" alt="screen shot 2018-11-07 at 19 21 17" src="https://user-images.githubusercontent.com/35170812/48144652-586ef680-e2c2-11e8-90c2-4607ea3d431e.png">


#### What are the relevant pivotal tracker stories?
[Finishes #161278918]
